docs/man.jpconj/man3/ConjFrame.3
.TH "ConjFrame" 3 "Tue Aug 29 2017" "Version 2.0.0" "JapKatsuyou.JpConj" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ConjFrame \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <conjframe\&.h>\fP
.PP
Inherits \fBFuncFrame\fP\&.
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBclose\fP ()"
.br
.ti -1c
.RI "void \fBopen\fP ()"
.br
.ti -1c
.RI "void \fBnewVerb\fP (QString verb)"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBConjFrame\fP (QWidget *parent=0)"
.br
.ti -1c
.RI "\fB~ConjFrame\fP ()"
.br
.ti -1c
.RI "void \fBdoConj\fP ()"
.br
.RI "\fI\fBConjFrame::doConj\fP This function is used to show different conjugation forms\&. \fP"
.ti -1c
.RI "virtual void \fBzoom\fP (signed char sign)"
.br
.ti -1c
.RI "virtual void \fBrefreshLanguage\fP (bool \fBrtl\fP)"
.br
.ti -1c
.RI "virtual void \fBchangeStyle\fP (QString styleID)"
.br
.ti -1c
.RI "virtual void \fBinitExporter\fP (\fBExport\fP *exporter)"
.br
.ti -1c
.RI "virtual void \fBchangeFont\fP (QString font, QString jpfont, int size, int jpsize)"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "virtual void \fBinitUI\fP ()"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Constructor & Destructor Documentation"
.PP 
.SS "ConjFrame::ConjFrame (QWidget * parent = \fC0\fP)\fC [explicit]\fP"

.SS "ConjFrame::~ConjFrame ()"

.SH "Member Function Documentation"
.PP 
.SS "void ConjFrame::changeFont (QString font, QString jpfont, int size, int jpsize)\fC [virtual]\fP"

.PP
Implements \fBFuncFrame\fP\&.
.SS "void ConjFrame::changeStyle (QString styleID)\fC [virtual]\fP"

.PP
Implements \fBFuncFrame\fP\&.
.SS "void ConjFrame::close ()\fC [signal]\fP"

.SS "void ConjFrame::doConj ()"

.PP
\fBConjFrame::doConj\fP This function is used to show different conjugation forms\&. This function search for the existance of the verb, using \fBEdict2\fP module\&. Then, it calls for:
.IP "\(bu" 2
ConjFrame::basicConjugation To show the standard and basic conjugation forms\&.
.IP "\(bu" 2
ConjFrame::complexConjugation To show the complex conjugation forms\&. 
.PP

.SS "void ConjFrame::initExporter (\fBExport\fP * exporter)\fC [virtual]\fP"

.PP
Implements \fBFuncFrame\fP\&.
.SS "void ConjFrame::initUI ()\fC [protected]\fP, \fC [virtual]\fP"

.PP
Implements \fBFuncFrame\fP\&.
.SS "void ConjFrame::newVerb (QString verb)\fC [signal]\fP"

.SS "void ConjFrame::open ()\fC [signal]\fP"

.SS "void ConjFrame::refreshLanguage (bool rtl)\fC [virtual]\fP"

.PP
Implements \fBFuncFrame\fP\&.
.SS "void ConjFrame::zoom (signed char sign)\fC [virtual]\fP"

.PP
Implements \fBFuncFrame\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for JapKatsuyou\&.JpConj from the source code\&.


CHANGELOG.md

# json-VCard Changelog

This is the changelog syntax:
* NEW: For any new module or file (creation)
* ADD: For any functions added inside a module (addition)
* IMPROVE: For any improvements to existing materials.
This affects the behavior, but doesn't break any precedent
* MODIFY: Modification of function names, files names and directories.
This will break the projects developed on precedent versions.
* FIX: Fixing bugs
* DEL: Delete or deprecate some functions

## Version 2.0.0 (2016-07-03)

* ADD: Control the font (size and type) for both default language and Japanese
* NEW: Choose the icons set
* NEW: Show examples in default language
* ADD: Show Furigana and Romaji pronunciation of the verb

## Version 1.0.1 (2014-01-01)

* I don't know what changed :)

## Version 0.2.0 (2013-09-04)

* NEW: first release


jpconj/Ui/conjframe.cpp
/*
    This file is part of JapKatsuyou project; an application that provide
    Japanese verb conjugation

    Copyright (C) 2013-2016  AnqaFalak group
    Copyright (C) 2013  Abdelkrime Aries <kariminfo0@gmail.com>

    Authors:
            Abdelkrime Aries <kariminfo0@gmail.com>

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
#include "conjframe.h"
#include "ui_conjframe.h"

ConjFrame::ConjFrame(QWidget *parent) :
    FuncFrame(parent),
    ui(new Ui::ConjFrame)
{
    initialize();
}

ConjFrame::~ConjFrame()
{
    delete ui;
}


void ConjFrame::initUI()
{
    ui->setupUi(this);
    QPalette palette;

    //hasContent = false;
    palette.setBrush(QPalette::Base, Qt::transparent);

    ui->verbInfo->page()->setPalette(palette);
    ui->verbInfo->setAttribute(Qt::WA_OpaquePaintEvent, false);

    ui->basicConj->page()->setPalette(palette);
    ui->basicConj->setAttribute(Qt::WA_OpaquePaintEvent, false);

    ui->standardConj->page()->setPalette(palette);
    ui->standardConj->setAttribute(Qt::WA_OpaquePaintEvent, false);

    ui->complexConj->page()->setPalette(palette);
    ui->complexConj->setAttribute(Qt::WA_OpaquePaintEvent, false);

    changeStyle(Style::getCurrentStyle());
}


void ConjFrame::initExporter(Export *exporter)
{
    exporter->setRTL(rtl);
    exporter->addContent("<p><h1>" + jaVerb.kanji + "</h1></p><hr>\n");
    //TODO add verb info here
    //exporter->addContent("<p><h3>" + ui->verbType->text() + "</h3></p>\n");
    if(Export::getConfigExportPart("info")){
        exporter->addContent("<p><h2>" + ui->ConjgTab->tabText(0) + "</h2></p>\n");
        QString data = ui->verbInfo->page()->mainFrame()->findFirstElement("body").firstChild().toOuterXml();
        exporter->addContent(data);
    }
    if(Export::getConfigExportPart("standard")){
        exporter->addContent("<p><h2>" + ui->ConjgTab->tabText(1) + "</h2></p>\n");
        QString data = ui->standardConj->page()->mainFrame()->findFirstElement("body").firstChild().toOuterXml();
        exporter->addContent(data);
    }
    if(Export::getConfigExportPart("basic")){
        exporter->addContent("<p><h2>" + ui->ConjgTab->tabText(2) + "</h2></p>\n");
        QString data = ui->basicConj->page()->mainFrame()->findFirstElement("body").firstChild().toOuterXml();
        exporter->addContent(data);
    }
    if(Export::getConfigExportPart("complex")){
        exporter->addContent("<p><h2>" + ui->ConjgTab->tabText(3) + "</h2></p>\n");
        QString data = ui->complexConj->page()->mainFrame()->findFirstElement("body").firstChild().toOuterXml();
        exporter->addContent(data);
    }

    if(Export::getConfigExportPart("styled")){
        QString styleCSS = readCSSFile(QDir(QString(dataFolder) + "styles/" + stylesheet).absolutePath());
        exporter->addStyle(styleCSS);
    }
}

void ConjFrame::changeFont(QString font, QString jpfont, int size, int jpsize)
{
    QString style = ".lan {font-family:" + font;
    style += "; font-size :" + QString::number(size) + "pt;}\n";

    style += ".jp {font-family:" + jpfont;
    style += "; font-size :" + QString::number(jpsize) + "pt;}";

    QWebElement infoEl = ui->verbInfo->page()->mainFrame()->findFirstElement("style");
    infoEl.setInnerXml(style);

    QWebElement basicEl = ui->basicConj->page()->mainFrame()->findFirstElement("style");
    basicEl.setInnerXml(style);

    QWebElement stdEl = ui->standardConj->page()->mainFrame()->findFirstElement("style");
    stdEl.setInnerXml(style);

    QWebElement cplxEl = ui->complexConj->page()->mainFrame()->findFirstElement("style");
    cplxEl.setInnerXml(style);

    //qDebug()<< "font changed" << ui->basicConj->page()->mainFrame()->toHtml();



}



/*!
 * \brief ConjFrame::doConj This function is used to show different conjugation forms
 *
 * This function search for the existance of the verb, using Edict2 module.
 * Then, it calls for:
 * - ConjFrame::basicConjugation To show the standard and basic conjugation forms.
 * - ConjFrame::complexConjugation To show the complex conjugation forms.
 */

void ConjFrame::doConj()
{

    QString verb = ui->inputConjVerb->text();

    if (verb == jaVerb.kanji)
        return;

    Edict2 edict2;

    Edict2::JaVerb javerb = edict2.find(verb);



    if (javerb.type < 1){
        //TODO show a message that the verb is not found
        //ui->verbType->setText(Msg::getVerbTypeDesc(type));
        close();
        javerb.kanji = "";
        javerb.hiragana = "";
        javerb.type = (EdictType) 0;
        hasContent = false;
        ui->verbInfo->setHtml("");
        ui->standardConj->setHtml("");
        ui->basicConj->setHtml("");
        ui->complexConj->setHtml("");

        return;
    }

    jaVerb = javerb;
    //ui->verbType->setText(Msg::getVerbTypeDesc(type));
    verbInformation(javerb);
    complexConjugation(javerb);
    basicConjugation(javerb);

    hasContent = true;

    refreshLanguage(rtl);
    newVerb(javerb.kanji);

    open();
}




/*!
 * \brief jpconjmain::basicConjugation Used to show the standard and basic conjugation forms.
 * \param verb The verb in dictionary form (u-form), eg. 食べる, 飲む, 行く, 来る, etc.
 * \param type The Edict2 type of the verb (See: VerbType::EdictType)
 */
void ConjFrame::basicConjugation(Edict2::JaVerb javerb)
{

    if (!hasContent){
        QString basicConjHTML = readHtmlFile(":/output/basicConj");
        QString standardConjHTML = readHtmlFile(":/output/standardConj");

        ui->standardConj->setHtml(standardConjHTML);
        ui->basicConj->setHtml(basicConjHTML);
    }

    QString verb = javerb.kanji;
    EdictType type = javerb.type;

    //! [Doxygen: basicFormsMap example]
    QMap<KForm, QString> basicForms = Msg::basicFormsMap();
    QWebElement element_stem;
    QWebElement element_suffix;
    QWebElement element_basic;
    foreach (KForm form, basicForms.keys()){
        QStringList conj = JpConj::Katsuyou(verb, type, form).split("|");
        QString elementId = basicForms.value(form);

        element_stem = ui->standardConj->page()->mainFrame()->findFirstElement("#stem_" + elementId);
        element_suffix = ui->standardConj->page()->mainFrame()->findFirstElement("#suffix_" + elementId);
        element_stem.setInnerXml(conj[0]);
        element_suffix.setInnerXml(conj[1]);

        element_basic = ui->basicConj->page()->mainFrame()->findFirstElement("#basic_" + elementId);
        element_basic.setInnerXml(conj[0] + conj[1]);

    }
    //! [Doxygen: basicFormsMap example]

}



/*!
 * \brief jpconjmain::complexConjugation Used to show the complex conjugation forms.
 * \param verb The verb in dictionary form (u-form), eg. 食, 飲む, 行く, 来る, etc.
 * \param type The Edict2 type of the verb (See: VerbType::EdictType)
 */
void ConjFrame::complexConjugation(Edict2::JaVerb javerb)
{
    if (!hasContent){
        QString complexConjHTML = readHtmlFile(":/output/complexConj");
        ui->complexConj->setHtml(complexConjHTML);
    }

    QString verb = javerb.kanji;
    EdictType type = javerb.type;

    QString jsScript="";
    QMap<CForm, QString> complexForms = Msg::complexFormsMap();
    foreach (CForm form, complexForms.keys()){

        QString elementId = complexForms.value(form);

        jsScript += "document.getElementById(\"PoA_" + elementId + "\").innerHTML = \"";
        jsScript += JpConj::Conjugate(verb, type, form, _Polite, _Affirmative).remove("|") + "\";\n";

        jsScript += "document.getElementById(\"PoN_" + elementId + "\").innerHTML = \"";
        jsScript += JpConj::Conjugate(verb, type, form, _Polite, _Negative).remove("|") + "\";\n";

        jsScript += "document.getElementById(\"PlA_" + elementId + "\").innerHTML = \"";
        jsScript += JpConj::Conjugate(verb, type, form, _Plain, _Affirmative).remove("|") + "\";\n";

        jsScript += "document.getElementById(\"PlN_" + elementId + "\").innerHTML = \"";
        jsScript += JpConj::Conjugate(verb, type, form, _Plain, _Negative).remove("|") + "\";\n";
        //complexConjHTML.replace("&_Form_" + str, Msg::getVerbFormName(form));
        //complexConjHTML.replace("&_Tip_" + str, Msg::getVerbFormDesc(form));
    }

    ui->complexConj->page()->mainFrame()->evaluateJavaScript(jsScript);

}

void ConjFrame::verbInformation(Edict2::JaVerb javerb)
{
    if (!hasContent){
        QString verbInfoHTML = readHtmlFile(":/output/verbInfo");
        ui->verbInfo->setHtml(verbInfoHTML);
    }

    QString verb = javerb.kanji;
    EdictType type = javerb.type;

    QString jsScript="";

    jsScript += "document.getElementById(\"_Verb\").innerHTML = \"";
    jsScript += Msg::getTranslatedString("_Verb") + "\";\n";

    jsScript += "document.getElementById(\"furigana\").innerHTML = \"";
    jsScript += "<ruby>" + verb + "<rt>" + javerb.hiragana + "</rt></ruby>\";\n";

    jsScript += "document.getElementById(\"_Romaji\").innerHTML = \"";
    jsScript += Msg::getTranslatedString("_Romaji") + "\";\n";

    jsScript += "document.getElementById(\"romaji\").innerHTML = \"";
    jsScript += JTrans::hiragana2romaji(javerb.hiragana) + "\";\n";

    jsScript += "document.getElementById(\"_Type\").innerHTML = \"";
    jsScript += Msg::getTranslatedString("_Type") + "\";\n";

    jsScript += "document.getElementById(\"type\").innerHTML = \"";
    jsScript += Msg::getVerbTypeDesc(type) + "\";\n";

    ui->verbInfo->page()->mainFrame()->evaluateJavaScript(jsScript);
}

/*!
 * \brief jpconjmain::readHtmlFile Reads an HTML file and return a QString
 * \param URL URL of the HTML file, we want to read.
 * \return A QString which is the content of this HTML file.
 */
QString ConjFrame::readHtmlFile(QString URL)
{
    QString result = "<!doctype html>\n";
    result += "<html>\n";
    result += "<meta http-equiv=\"content-type\" content=\"text/html; charset=UTF-8\" />\n";
    result += "<style>\n";
    result += ".lan {font-family:" + Style::getCurrentFont(false) + ";";
    result += "font-size:" + QString::number(Style::getCurrentFontSize(false)) + "pt;}\n";
    result += ".jp {font-family:" + Style::getCurrentFont(true);
    result += "; font-size :" + QString::number(Style::getCurrentFontSize(true)) + "pt;}\n";
    result += "\n</style>\n";
    result += "<body>\n";

    QFile HtmlFile(URL);

    if (HtmlFile.open(QIODevice::ReadOnly | QIODevice::Text)){
        QTextStream htmlStream(&HtmlFile);
        result += htmlStream.readAll();
    }

    result += "</body>\n</html>";

    return result;
}

QString ConjFrame::readCSSFile(QString URL)
{
    QString result="";
    QFile HtmlFile(URL);

    if (HtmlFile.open(QIODevice::ReadOnly | QIODevice::Text)){
        QTextStream htmlStream(&HtmlFile);
        result = htmlStream.readAll();
    }

    return result;
}


void ConjFrame::refreshLanguage(bool rtl)
{

    this->rtl = rtl;

    ui->retranslateUi(this);

    if (!hasContent)
        return;

    QString jsScript = "var body = document.getElementsByTagName('body')[0]; \n";
    QString dir = (rtl)?"rtl":"ltr";
    jsScript += "body.dir = \"" + dir + "\";";

    //ui->standardConj->page()->mainFrame()->evaluateJavaScript(jsScript);
    //it is better for standard conjugation to stay ltr
    ui->basicConj->page()->mainFrame()->evaluateJavaScript(jsScript);
    ui->complexConj->page()->mainFrame()->evaluateJavaScript(jsScript);
    ui->verbInfo->page()->mainFrame()->evaluateJavaScript(jsScript);
    {
        QWebElement element = ui->verbInfo->page()->mainFrame()->findFirstElement("#type");
        element.setAttribute("dir", dir);
    }
    //qDebug()<< jsScript;

    //Retranslate strings

    {//Information
        jsScript = "document.getElementById(\"_Verb\").innerHTML = \"";
        jsScript += Msg::getTranslatedString("_Verb") + "\";\n";
        jsScript += "document.getElementById(\"_Romaji\").innerHTML = \"";
        jsScript += Msg::getTranslatedString("_Romaji") + "\";\n";
        jsScript += "document.getElementById(\"_Type\").innerHTML = \"";
        jsScript += Msg::getTranslatedString("_Type") + "\";\n";
        //ui->verbInfo->page()->mainFrame()->evaluateJavaScript(jsScript);
        jsScript += "document.getElementById(\"type\").innerHTML = \"";
        jsScript += Msg::getVerbTypeDesc(jaVerb.type) + "\";\n";
        ui->verbInfo->page()->mainFrame()->evaluateJavaScript(jsScript);
    }

    {//standard
        QWebElementCollection standardConjConst = ui->standardConj->page()->mainFrame()->findAllElements(".Const");

        for(int i = 0; i < standardConjConst.count(); i++){
            QWebElement element = standardConjConst.at(i);
            QString elementName = element.attribute("name", "");
            element.setInnerXml(Msg::getTranslatedString(elementName));
        }
    }

    {//basic
        QWebElementCollection basicConjConst = ui->basicConj->page()->mainFrame()->findAllElements(".Const");

        for(int i = 0; i < basicConjConst.count(); i++){
            QWebElement element = basicConjConst.at(i);
            QString elementName = element.attribute("name", "");
            element.setInnerXml(Msg::getTranslatedString(elementName));
        }

        QMap<KForm, QString> basicForms = Msg::basicFormsMap();
        QWebElement element_basic;
        foreach (KForm form, basicForms.keys()){
            QString elementId = basicForms.value(form);
            element_basic = ui->basicConj->page()->mainFrame()->findFirstElement("#_Name_" + elementId);
            element_basic.setInnerXml(Msg::getBasicFormName(form));
        }
    }

    //complexConj

    QWebElementCollection complexConjConst = ui->complexConj->page()->mainFrame()->findAllElements("[name=\"_Form\"]");
    for(int i = 0; i < complexConjConst.count(); i++)
        complexConjConst.at(i).setInnerXml(Msg::getTranslatedString("_Form"));

    complexConjConst = ui->complexConj->page()->mainFrame()->findAllElements("[name=\"_Polite\"]");
    for(int i = 0; i < complexConjConst.count(); i++){
        QWebElement element = complexConjConst.at(i);
        element.setInnerXml(Msg::getVerbPolitenessName(_Polite));
        element.setAttribute("title", Msg::getVerbPolitenessDesc(_Polite));
    }

    complexConjConst = ui->complexConj->page()->mainFrame()->findAllElements("[name=\"_Plain\"]");
    for(int i = 0; i < complexConjConst.count(); i++){
        QWebElement element = complexConjConst.at(i);
        element.setInnerXml(Msg::getVerbPolitenessName(_Plain));
        element.setAttribute("title", Msg::getVerbPolitenessDesc(_Plain));
    }

    complexConjConst = ui->complexConj->page()->mainFrame()->findAllElements("[name=\"_Affirmative\"]");
    for(int i = 0; i < complexConjConst.count(); i++){
        QWebElement element = complexConjConst.at(i);
        element.setInnerXml(Msg::getVerbPolarityName(_Affirmative));
        element.setAttribute("title", Msg::getVerbPolarityDesc(_Affirmative));
    }

    complexConjConst = ui->complexConj->page()->mainFrame()->findAllElements("[name=\"_Negative\"]");
    for(int i = 0; i < complexConjConst.count(); i++){
        QWebElement element = complexConjConst.at(i);
        element.setInnerXml(Msg::getVerbPolarityName(_Negative));
        element.setAttribute("title", Msg::getVerbPolarityDesc(_Negative));
    }

    QMap<CForm, QString> complexForms = Msg::complexFormsMap();
    QWebElement element_complex;
    foreach (CForm form, complexForms.keys()){
        QString elementId = complexForms.value(form);

        element_complex = ui->complexConj->page()->mainFrame()->findFirstElement("#_Form_" + elementId);
        element_complex.setInnerXml(Msg::getVerbFormName(form));
        element_complex.setAttribute("title", Msg::getVerbFormDesc(form));
    }
    //qDebug()<< jsScript;
    ui->complexConj->page()->mainFrame()->evaluateJavaScript(jsScript);

    //qDebug()<< "Strings translation";

}



void ConjFrame::zoom(signed char sign)
{
    if (sign < 0){
        ui->standardConj->setTextSizeMultiplier(qMax(0.5, ui->standardConj->textSizeMultiplier() - 1.0 / 10.0));
        ui->basicConj->setTextSizeMultiplier(qMax(0.5, ui->basicConj->textSizeMultiplier() - 1.0 / 10.0));
        ui->complexConj->setTextSizeMultiplier(qMax(0.5, ui->complexConj->textSizeMultiplier() - 1.0 / 10.0));

        return;
    }

    if (sign > 0){
        ui->standardConj->setTextSizeMultiplier(qMin(2.5,ui->standardConj->textSizeMultiplier() + 1.0 / 10.0));
        ui->basicConj->setTextSizeMultiplier(qMin(2.5, ui->basicConj->textSizeMultiplier() + 1.0 / 10.0));
        ui->complexConj->setTextSizeMultiplier(qMin(2.5, ui->complexConj->textSizeMultiplier() + 1.0 / 10.0));

        return;
    }

    ui->standardConj->setTextSizeMultiplier(1.0);
    ui->basicConj->setTextSizeMultiplier(1.0);
    ui->complexConj->setTextSizeMultiplier(1.0);
}


void ConjFrame::changeStyle(QString styleID)
{
    stylesheet = styleID + ".css";
    //qDebug() << "style changed" << styleID;
    setCSS(ui->basicConj, stylesheet);
    setCSS(ui->standardConj, stylesheet);
    setCSS(ui->complexConj, stylesheet);
    setCSS(ui->verbInfo, stylesheet);
}


void ConjFrame::on_conjugateButton_clicked()
{
    doConj();
}


void ConjFrame::on_inputConjVerb_returnPressed()
{
    doConj();
}
